include/wil/winrt.h

```diff
@@ -66,1216 +66,1270 @@ namespace std
+//! map.find(constArray);
+//! map.find(nonConstArray);
+//! map.find(static_cast<const wchar_t*>(constArray));
+//! ~~~
+//! The first four calls in the example above use `WindowsGetStringRawBuffer` (or equivalent) to get the string
+//! buffer and length for the comparison. The fifth example uses `std::wstring::c_str` and `std::wstring::length`
+//! for getting these two values. The remaining three examples use only the string buffer and call `wcslen` for the
+//! length. That is, the length is *not* deduced for either array. This is because argument types are not always
+//! perfectly preserved by container functions and in fact are often captured as const references making it
+//! impossible to differentiate const arrays - where we can safely deduce length - from non const arrays - where we
+//! cannot safely deduce length since the buffer may be larger than actually needed (e.g. creating a
+//! `char[MAX_PATH]` array, but only filling it with 10 characters). The implications of this behavior is that
+//! string literals that contain embedded null characters will only include the part of the buffer up to the first
+//! null character. For example, the following example will result in all calls to `find` returning an end
+//! iterator.
+//! ~~~
+//! std::map<HString, int, wil::hstring_less> map;
+//! const wchar_t constArray[] = L"foo\0bar";
+//! wchar_t nonConstArray[MAX_PATH] = L"foo\0bar";
+//!
+//! // Create the key with the embedded null character
+//! HString key;
+//! THROW_IF_FAILED(key.Set(constArray));
+//! map.emplace(std::move(key), 42);
+//!
+//! // All of the following return map.end() since they look for the string "foo"
+//! map.find(constArray);
+//! map.find(nonConstArray);
+//! map.find(static_cast<const wchar_t*>(constArray));
+//! ~~~
+//! In order to search using a string literal that contains embedded null characters, a simple alternative is to
+//! first create an `HStringReference` and use that for the function call:
+//! ~~~
+//! // HStringReference's constructor *will* deduce the length of const arrays
+//! map.find(HStringReference(constArray));
+//! ~~~
+struct hstring_less
+{
+    using is_transparent = void;
 
-        class vector_iterator
-        {
-        public:
-#ifdef _XUTILITY_
-            // could be random_access_iterator_tag but missing some features
-            typedef ::std::bidirectional_iterator_tag iterator_category;
-#endif
-            typedef TSmart value_type;
-            typedef ptrdiff_t difference_type;
-            typedef const TSmart* pointer;
-            typedef const TSmart& reference;
-
-            // for begin()
-            vector_iterator(VectorType* v, unsigned int pos)
-                : m_v(v), m_i(pos)
-            {
-            }
+    template <typename LhsT, typename RhsT>
+    WI_NODISCARD auto operator()(const LhsT& lhs, const RhsT& rhs) const
+        WI_NOEXCEPT->decltype(details::hstring_compare<true, false>::less(lhs, rhs))

```

Comment:
Interesting behavior: looks like it's treating `WI_NOEXCEPT` as a variable, so it seems to be assuming that `->` is doing pointer access

Reply:
Looks like `clang-format` version 17 added a `Macros` key that allows specifying mapping of definition to expansion, which would presumably fix this. Sadly it looks like VS ships with version 16 currently :(

Reply:
Tested it out and while I can confirm that it fixes cases like these, for some reason it also causes more aggressive line wrapping that doesn't occur without the `Macros` key, nor does it occur with the `noexcept` keyword itself. Seems like a bug in `clang-format`

Reply:
Using `Macros` also fixes the weird formatting of member initializers on a `noexcept` constructor